extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +151,76 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticSnapshotsMap = new Map<string, readonly vscode.Diagnostic[]>();
+	private _timeout: NodeJS.Timeout | undefined;
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _getDiagnostics: (uri: vscode.Uri) => readonly vscode.Diagnostic[]
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript') {
+				this._updateAllDiagnosticCodesAfterTimeout();
+			}
+		}));
+		this._updateAllDiagnosticCodesAfterTimeout();
+		this._registerTelemetryEventEmitter();
+	}
+
+	private _updateAllDiagnosticCodesAfterTimeout() {
+		clearTimeout(this._timeout);
+		const uris = vscode.workspace.textDocuments.map(doc => doc.uri);
+		this._timeout = setTimeout(() => { uris.forEach((uri) => this._updateDiagnosticCodes(uri)); }, 5000);
+	}
+
+	private _increaseDiagnosticCodeCount(code: string | number | undefined) {
+		if (code === undefined) {
+			return;
+		}
+		this._diagnosticCodesMap.set(Number(code), (this._diagnosticCodesMap.get(Number(code)) || 0) + 1);
+	}
+
+	private _updateDiagnosticCodes(uri: vscode.Uri) {
```

Comment:
As it is today, this might leak diagnostics because sometimes diagnostics for closed files are fixed by typing in an open file. `vscode.workspace.textDocuments` only returns open documents.

To tackle that, we could remove `uri` from here. We could pass to this class the entire `vscode.DiagnosticCollection` which allows to loop over all TS diagnostics across all files.